Bluetooth: Mesh: Add model publish support to Health Client

michal-narajowski opened the pull request:
Mesh Profile Specification errata 11310
Section 4.4.4.1
"This model shall support model publication, as defined in Section
4.2.2 of the Mesh Profile specification and model subscription,
as defined in section 4.2.3 of the Mesh Profile specification."

Signed-off-by: Michal Narajowski <michal.narajowski@codecoup.pl>

PavelVPV commented:
@michal-narajowski could you add errata number to the commit message?

PavelVPV commented:
As was discussed some time ago, this is not enough since health client builds its own message context in its api functions leaving publication context unused.

PavelVPV commented:
I see 2 ways to support model publication in health client:

1. Introduce a separate API just for sending messages using model publication. E.g. for `bt_mesh_health_fault_get` add `bt_mesh_health_fault_get_pub`, for `bt_mesh_health_fault_clear_unack` add `bt_mesh_health_fault_clear_unack_pub`, etc. Thus we don't break current API, don't need to deprecate it either.

2. Mark the current API as deprecated, establish a new one that will support both, sending a message using model publication and direct sending. The new API e.g. has `bt_mesh_health_cli_` prefix and takes health client and msg ctx as arguments like you did.

I would prefer the second way as it would not create an overhead in the API. @jhedberg @LingaoM @omkar3141 what do you think?

omkar3141 commented:
> I see 2 ways to support model publication in health client:
> 
> 1. Introduce a separate API just for sending messages using model publication. E.g. for `bt_mesh_health_fault_get` add `bt_mesh_health_fault_get_pub`, for `bt_mesh_health_fault_clear_unack` add `bt_mesh_health_fault_clear_unack_pub`, etc. Thus we don't break current API, don't need to deprecate it either.
> 2. Mark the current API as deprecated, establish a new one that will support both, sending a message using model publication and direct sending. The new API e.g. has `bt_mesh_health_cli_` prefix and takes health client and msg ctx as arguments like you did.
> 
> I would prefer the second way as it would not create an overhead in the API. @jhedberg @LingaoM @omkar3141 what do you think?

I would prefer the second option to align all client model APIs.

jhedberg reviewed (changes requested):
I'd need to get much more info in this issue. The erratum referenced doesn't really help explain how publishing is supposed to work with the model. I tried to download the latest MeshPrf 1.1 draft spec to get some idea of this, but so far I didn't manage to find the relevant info (e.g. to which messages publication applies, how does this work with periodic publication, etc- I can try infer it from the content of this PR, but I'd prefer to see the spec text itself). 

Btw, based on the erratum it seems like this is anyway for Mesh 1.1, i.e. not yet ready upstream inclusion until that spec version gets adopted?

Is the core issue that the existing model publication support in the Zephyr Mesh stack has been implemented with the assumption that there is a single message that's used for publication, and the current `bt_mesh_model_publish()` API and single callback don't really adapt well to a situation where a model might use the model publication state for multiple different messages (as well as some messages involving an acknowledgement response)? If so, then I think we should take a step back and look at this from a more generic API level than anything Health client model specific.

Since I don't know all the answers to the above questions I can't really propose any solutions yet, but some thoughts come to mind:

 - Do we need some extended `bt_mesh_model_publish()` API that takes more parameters, e.g. a message buffer?
 - Would it be helpful to have an API that initializes a `bt_mesh_msg_ctx` struct from a model publication struct? E.g. `bt_mesh_msg_ctx_init_pub(struct bt_mesh_msg_ctx *ctx, struct bt_mesh_model_pub *pub);`?

omkar3141 commented:
All general-purpose clients (such as lighting model clients, or generic clients) support publishing. They do not specify a message to publish because it is envisioned that for the clients, the specific message depends on the end-product. For example, a certain light-switch by a certain manufacturer implementing a Generic OnOff client will always send out "Generic OnOff Set Unack" upon physical button press by the user. As long as publish callback is provided, the user application can call whichever client model API it wants as per the end-product needs. That is the idea.

jhedberg commented:
@omkar3141 thanks for the clarification! However, that really only makes sense if the different message options are somewhat similar (e.g. same action, but one is asked and other snacked). I don't see how it'd make sense for the same model, but from different manufacturers, to send a completely different type of message, e.g. one sends a "get faults" message when you enable publishing, but the other sends a "clear faults" message. There's no way to make a cross-vendor interoperable network that way, since knowing the model in question still doesn't fully tell you how the model behaves.

jhedberg commented:
Btw, in my previous comment I'm mainly thinking about periodic publishing, i.e. message sending triggered by a timer, rather than an explicit action on the client side.

PavelVPV commented:
> > Also, the model publication state is defined per model, not per message. If a user can't control it for a specific message, it doesn't make sense to have a publication buffer per each message.
> 
> It does make sense to have a buffer if you assume that the parameters (payload) keep changing, but the opcode always remains the same for the same model (this is how I originally had understood publication when I implemented the Zephyr Mesh stack many years ago, but I now see that this was a two narrow definition of publication).

The buffer can be changed in the update callback. This is how it is used in conjunction with `retr_update` flag to support changing delay value in Generic Delta message.

jhedberg commented:
I took a more closer look at how this has been now implemented, and it's actually not that bad, since you *are* using the `bt_mesh_model_publish()` API (I first thought you were just picking the meta-data from the publication struct and using other message sending APIs with it).

One concern I have with the deprecation option: if we use the `bt_mesh_health_cli_*` prefix this becomes inconsistent with the configuration client API naming (which doesn't have "cli" in it). We can either ignore this concern, use the stable API process (i.e. break the API backwards compatibility) or to change the configuration client API as well.

My other question still remains though: do you see any ways in which we could improve the generic (non model specific) model publication API to make it easier to be used for this kind of use cases?

jhedberg commented:
Oh, and still one other unanswered question: since the erratum is for Mesh 1.1, do we anyway want to merge this change upstream right now, since the spec version supported is 1.0.1?

PavelVPV commented:
> One concern I have with the deprecation option: if we use the `bt_mesh_health_cli_*` prefix this becomes inconsistent with the configuration client API naming (which doesn't have "cli" in it). We can either ignore this concern, use the stable API process (i.e. break the API backwards compatibility) or to change the configuration client API as well.

From the nRF Connect SDK experience, I would see beneficial of adding `_cli` infix into every client model API. This helps easily distinguish clients API from server one which could have `_srv` infix. This means that `bt_mesh_fault_update` also has to be renamed to `bt_mesh_health_srv_fault_update`. This will also be future-proof for mesh 1.1 models, which may some API for server side.

> My other question still remains though: do you see any ways in which we could improve the generic (non model specific) model publication API to make it easier to be used for this kind of use cases?

What we do in nRF Connect SDK is that we memcpy a new message to the publication buffer every time before calling `bt_mesh_model_publish` if we need to publish a new message (https://github.com/nrfconnect/sdk-nrf/blob/535c7a07689044cc54bc242a4a74014068eee886/subsys/bluetooth/mesh/model_utils.c#L116-L119), which looks like a trick. So it would be a good improvement. One downside could be in this is that if someone wants to trigger publication without changing the buffer. Then converting `bt_mesh_model_publish(void)` to something like `bt_mesh_model_publish(struct net_buf_simple *buf)` will create the opposite issue. But I don't see where it can be needed.

michal-narajowski commented:
> Oh, and still one other unanswered question: since the erratum is for Mesh 1.1, do we anyway want to merge this change upstream right now, since the spec version supported is 1.0.1?

This erratum is actually for Mesh Profile 1.0 but was incorporated in Mesh Profile draft 1.1 revision 00. At least this is how I understand it. Is this right, @omkar3141?

Commit:
Bluetooth: Mesh: Add model publish support to Health Client

Mesh Profile Specification errata 11310
Section 4.4.4.1
"This model shall support model publication, as defined in Section
4.2.2 of the Mesh Profile specification and model subscription,
as defined in section 4.2.3 of the Mesh Profile specification."

Signed-off-by: Michal Narajowski <michal.narajowski@codecoup.pl>

Commit:
tests: bluetooth: tester: Fix Health Fault Test command

The buffer for response was too small.

Signed-off-by: Michal Narajowski <michal.narajowski@codecoup.pl>

michal-narajowski commented:
@omkar @PavelVPV @jhedberg I updated the PR by leaving the old API and adding new one. I also added new server api to match the client. I don't know how to mark the old API as deprecated. Is this change ok? After that I can also align cfg_cli API if needed and update usages in apps and tests.

PavelVPV reviewed (changes requested):
IMO this looks good at the moment. Regarding deprecation, the deprecated API needs to have `__deprecated` in definition (https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated).

I would like the current API to be deprecated rather than breaking compatibility as this would bring less hassle for everyone.

michal-narajowski commented on include/zephyr/bluetooth/mesh/health_cli.h:
```diff
@@ -322,6 +331,229 @@ int bt_mesh_health_attention_set(uint16_t addr, uint16_t app_idx,
+ *
+ *  @return 0 on success, or (negative) error code on failure.
+ */
+int bt_mesh_health_cli_fault_get(struct bt_mesh_health_cli *cli, struct bt_mesh_msg_ctx *ctx,
```
Mesh Profile Specification v1.0.1, Section 4.4.4.1:
```
If supported, the Health Client model shall be supported by a primary element and may be supported by
any secondary elements. 
```
So to me this means that there can be multiple client models.

PavelVPV replied:
Right!

michal-narajowski commented on include/zephyr/bluetooth/mesh/health_cli.h:
```diff
@@ -322,6 +331,229 @@ int bt_mesh_health_attention_set(uint16_t addr, uint16_t app_idx,
+ *  @return 0 on success, or (negative) error code on failure.
+ */
+int bt_mesh_health_cli_fault_get(struct bt_mesh_health_cli *cli, struct bt_mesh_msg_ctx *ctx,
+				 uint16_t cid, uint8_t *test_id, uint8_t *faults,
```
It is possible. Personally, I don't think there is too many arguments now, but I can change it if there is a different conclusion.

PavelVPV replied:
It could be good to have less than or equal to 4 arguments in the function so that they are passed through registers.

michal-narajowski replied:
Updated.

PavelVPV replied:
Ah, in its current form, this structure only introduces more ambiguity as it is unclear in which case test id needs to be passed and in which it is returned back. Converting this structure to `struct health_fault_param` also seems useless. I tried something like this:

```
struct bt_mesh_health_cli_fault_rsp {
        /* Test ID response buffer. */
        uint8_t test_id;
        /* Fault array response buffer. */
        uint8_t *faults;
        /* Fault count response buffer. */
        size_t fault_count;
};

struct bt_mesh_health_cli_fault_params {
        /*  Company ID to get the registered faults of. */
        uint16_t cid;
        /* Identifier of a test to be performed. */
        uint8_t *test_id;
        /* Fault Status response. */
        struct bt_mesh_health_cli_fault_rsp *rsp;
};
```

with removing `struct health_fault_param`, but it doesn't bring any benefit either. So I would suggest to rollback this change. Sorry for misleading you.

michal-narajowski replied:
I reverted the last change.

PavelVPV reviewed (commented):
Looks good. One last minor comment. Could you update health client documentation?
